site/src/hooks/usePaginatedQuery.ts

```diff
@@ -0,0 +1,443 @@
+  const currentPage = parsePage(searchParams);
+  const currentPageOffset = (currentPage - 1) * limit;
+
+  const getQueryOptionsFromPage = (pageNumber: number) => {
+    const pageParams: QueryPageParams = {
+      pageNumber,
+      limit,
+      offset: (pageNumber - 1) * limit,
+      searchParams: getParamsWithoutPage(searchParams),
+    };
+
+    const payload = queryPayload?.(pageParams) as RuntimePayload<TQueryPayload>;
+
+    return {
+      queryKey: queryKey({ ...pageParams, payload }),
+      queryFn: (context: QueryFunctionContext<TQueryKey>) => {
+        return outerQueryFn({ ...context, ...pageParams, payload });
+      },
+    } as const;
+  };
+
+  // Not using infinite query right now because that requires a fair bit of list
+  // virtualization as the lists get bigger (especially for the audit logs).
+  // Keeping initial implementation simple.
+  const query = useQuery<TQueryFnData, TError, TData, TQueryKey>({
+    ...extraOptions,
+    ...getQueryOptionsFromPage(currentPage),
+    keepPreviousData: true,
+  });
+
+  const totalRecords = query.data?.count;
+  const totalPages =
+    totalRecords !== undefined ? Math.ceil(totalRecords / limit) : undefined;
+
+  const hasNextPage =
+    totalRecords !== undefined && limit + currentPageOffset < totalRecords;
+  const hasPreviousPage =
+    totalRecords !== undefined &&
+    currentPage > 1 &&
+    currentPageOffset - limit < totalRecords;
+
+  const queryClient = useQueryClient();
+  const prefetchPage = useEffectEvent((newPage: number) => {
+    const options = getQueryOptionsFromPage(newPage);
+    return queryClient.prefetchQuery(options);
+  });
+
+  // Have to split hairs and sync on both the current page and the hasXPage
+  // variables, because the page can change immediately client-side, but the
+  // hasXPage values are derived from the server and won't always be immediately
+  // ready on the initial render
+  useEffect(() => {
+    if (hasNextPage) {
+      void prefetchPage(currentPage + 1);
+    }
+  }, [prefetchPage, currentPage, hasNextPage]);
+
+  useEffect(() => {
+    if (hasPreviousPage) {
+      void prefetchPage(currentPage - 1);
+    }
+  }, [prefetchPage, currentPage, hasPreviousPage]);
+
+  // Mainly here to catch user if they navigate to a page directly via URL;
+  // totalPages parameterized to insulate function from fetch status changes
+  const updatePageIfInvalid = useEffectEvent(async (totalPages: number) => {
+    // If totalPages is 0, that's a sign that the currentPage overshot, and the
+    // API returned a count of 0 because it didn't know how to process the query
+    let fixedTotalPages: number;
+    if (totalPages !== 0) {
+      fixedTotalPages = totalPages;
+    } else {
+      const firstPageOptions = getQueryOptionsFromPage(1);
+      try {
+        const firstPageResult = await queryClient.fetchQuery(firstPageOptions);
+        fixedTotalPages = Math.ceil(firstPageResult?.count ?? 0 / limit) || 1;
+      } catch (err) {
```

Comment:
```suggestion
      } catch {
```

no need to capture the error as a variable

Reply:
Oh, I had no idea that you could have a catch block without the error definition. I'll try not to abuse that too much